### PR TITLE
handle bind request in adapter

### PR DESCRIPTION
**Pull request by cpacejo**
Also:
* relocate mgmt handling stuff into mgmt module
* thread ingress link ID through mgmt code
* eliminate most panics from mgmt handling path

### CODE REVIEW

File: adapter/ph/src/mgmt.rs [hunk 1]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+            HandleMgmtError::UnknownType(_type) => Self::UnknownType,
+            HandleMgmtError::UnexpectedMgmtResponse => Self::UnexpectedMgmtResponse,
+            HandleMgmtError::BadStructure => Self::BadStructure,
+        }
+    }
+}
+
+pub type HandleMgmtResult<'pktbuf> = Result<(), (HandleMgmtError, Packet<'pktbuf>)>;
+
+pub async fn handle_report<'pktbuf>(
```
Comment (cpacejo) [inline adapter/ph/src/mgmt.rs:106]: These all were just moved, had panics changed to error returns, and had `ingress_link_id` threaded through.

File: adapter/ph/src/mgmt.rs [hunk 2]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+        zdp::ZdpPacketType::HelloResponse,
+        rsp_pkt,
+    )
+    .await;
+
+    Ok(())
+}
+
+// RFC 6.5 § 6.3.11
+pub async fn handle_bind_agent_address_request<'pktbuf>(
```
Comment (cpacejo) [inline adapter/ph/src/mgmt.rs:164]: This is the only method here which is actually new.

File: adapter/ph/src/mgmt_processor_worker.rs [hunk 3]
```diff
@@ -46,62 +54,39 @@ async fn handle_packet<'pktbuf>(asm: &Assembly<'pktbuf>, mut pkt: Packet<'pktbuf
...
+
+            ZdpPacketType::BindAgentAddressRequest => {
+                mgmt::handle_bind_agent_address_request(asm, ingress_link_id, stream_id, pkt).await
+            }
+
+            packet_type => Err((HandleMgmtError::UnknownType(packet_type.0), pkt)),
         }
     } else {
         match base_hdr.packet_type {
-            ZdpPacketType::Report => {
```
Comment (cpacejo) [inline adapter/ph/src/mgmt_processor_worker.rs:72]: These all got moved into the mgmt module.

File: zpr-ext/src/zerocopy.rs [hunk 4]
```diff
@@ -28,14 +32,29 @@ mod zerocopy_bytes {
...
+        }
+
         fn read_from_buf(buf: &mut impl Buf) -> Option<Self>
         where
             Self: Sized,
         {
             if buf.remaining() < std::mem::size_of::<Self>() {
                 None
             } else {
-                Self::read_from(buf.copy_to_bytes(std::mem::size_of::<Self>()).as_ref())
```
Comment (cpacejo) [inline zpr-ext/src/zerocopy.rs]: Drive-by premature optimization.  I realized `copy_to_bytes()` is performing a heap allocation.  So instead I rewrote this to allocate space for a `T` on the stack.  But that was very roundabout because of how the `zerocopy` traits are defined.

File: adapter/ph/src/mgmt.rs [hunk 5]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+    let Some(hdr) = zdp::ZdpReportHeader::ref_from_prefix(pkt.body()) else {
+        return Err((HandleMgmtError::BadStructure, pkt));
+    };
+
+    // TODO handle protocol errors i.e. if the body is shorter
+    let report_data_length: usize = hdr.report_data_length.into();
+    pkt.advance(std::mem::size_of::<zdp::ZdpReportHeader>());
+    if pkt.body().len() >= report_data_length {
+        // TODO printing to stderr blocks indefinitely, this is just temporary
+        eprintln!(
```
Comment (mkolehmainen) [inline adapter/ph/src/mgmt.rs:120]: Why not use tracing? Eg, `info!`.

File: adapter/ph/src/mgmt.rs [hunk 6]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+    if compression_mode & zpr::compression_mode::SOURCE_PORT_PRESENT != 0 {
+        if pkt.remaining() < 2 {
+            return Err((HandleMgmtError::BadStructure, pkt));
+        }
+        src_port = pkt.get_u16();
+    }
+
+    // read destination port (optional)
+    let mut dst_port = 0;
+    if compression_mode & zpr::compression_mode::SOURCE_PORT_PRESENT != 0 {
```
Comment (mkolehmainen) [inline adapter/ph/src/mgmt.rs:230]: Does `SOURCE_PORT_PRESENT` imply a destination port too?

File: adapter/ph/src/mgmt.rs [hunk 7]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+
+    let ip_version = hdr.ip_version;
+    let compression_mode = hdr.compression_mode;
+
+    // read addresses (always present)
+    let src_address;
+    let dst_address;
+    match ip_version {
+        zpr::L3Type::Ipv4 => {
+            let Some(src_addr) = <[u8; 4]>::read_from_buf(&mut pkt) else {
```
Comment (mkolehmainen) [inline adapter/ph/src/mgmt.rs:185]: Shouldn't we have a header or message struct we can read into instead of reading bytes ourselves?  Wouldn't it be nicer if the addresses were actually ip address types?

File: adapter/ph/src/mgmt.rs [hunk 8]
```diff
@@ -79,3 +84,213 @@ pub async fn send_discard<'pktbuf>(asm: &'pktbuf Assembly<'pktbuf>, link_id: zpr
...
+    // recycle request buffer for response
+    let mut rsp_pkt = Packet::new(pkt.destroy(), config::DEFAULT_MESSAGE_HEADROOM);
+
+    // attempt to insert into DLT
+    let stream_id;
+    match asm.dlt.insert(pep) {
+        Ok(sid) => {
+            // success; respond with stream ID
+            // TODO: maybe tick a counter somewhere?
+            zdp::ZdpBindAgentAddressResponseHeader {
```
Comment (mkolehmainen) [inline adapter/ph/src/mgmt.rs:259]: Haven't looked at spec, but feels like there should be a sequence number or something here that ties the response to the request.

File: adapter/ph/src/peer_table.rs [hunk 9]
```diff
@@ -7,13 +7,23 @@ use std::sync::Mutex;
...
+
 pub struct PeerState {
+    pub peer_type: PeerType,
     pub substrate_addr: SubstrateAddr,
 }
 
 impl PeerState {
-    pub fn new(substrate_addr: SubstrateAddr) -> Self {
-        Self { substrate_addr }
+    pub fn new(peer_type: PeerType, substrate_addr: SubstrateAddr) -> Self {
```
Comment (mkolehmainen) [inline adapter/ph/src/peer_table.rs:22]: In the prototype we use separate ports for dock and link connections.  Are we doing that in ph?  If not do we need an unknown type or should we use an Option?

File: adapter/ph/src/mgmt.rs (hunk 5 above)
Comment (cpacejo) [reply adapter/ph/src/mgmt.rs:120]: I agree, not new code though.

**Review by mkolehmainen — APPROVED** [review]
I have some questions/comments for you, but looks good.

File: adapter/ph/src/mgmt.rs (hunk 6 above)
Comment (cpacejo) [reply adapter/ph/src/mgmt.rs:230]: No, § 6.3.11 of RFC 6 has them as separate.

File: adapter/ph/src/mgmt.rs (hunk 7 above)
Comment (cpacejo) [reply adapter/ph/src/mgmt.rs:185]: Yeah this bugged me too but we already have a proliferation of "IP address" types.  (Rust's `Ipv4Addr`, `Ipv6Addr`, `IpAddr`, and our own `IpAddress`.)  Maybe I'll define `Ipv4AddressWire` and `Ipv4AddressWire` types which are exactly these wire representations though.

File: adapter/ph/src/mgmt.rs (hunk 6 above)
Comment (mkolehmainen) [reply adapter/ph/src/mgmt.rs:230]: But it's ok that you are using the same compression mode flag to read both src & dst ports?  `SOURCE_PORT_PRESENT`. ?

File: adapter/ph/src/mgmt.rs (hunk 8 above)
Comment (cpacejo) [reply adapter/ph/src/mgmt.rs:259]: Ohhhhhhhhhhhhhhhhhhhhhhhhhhhhhh is there some baggage here :)

(Quick summary: we have too many open questions about how sequence numbers are supposed to work.  We need to get Frank in to sort it all out before we can attempt to implement something.)

File: adapter/ph/src/peer_table.rs (hunk 9 above)
Comment (cpacejo) [reply adapter/ph/src/peer_table.rs:22]: I'm not following, which ports are you referring to?

The `SubstrateAddr` here refers to the remote address:port of the link or docking session.
